substrate/bin/utils/chain-spec-builder/src/lib.rs

```diff
@@ -279,18 +294,130 @@ pub struct VerifyCmd {
+				ref input_chain_spec,
+				ref runtime,
+				block_height,
+			}) => {
+				let mut chain_spec_json = extract_chain_spec_json(input_chain_spec.as_path())?;
+
+				set_code_substitute_in_json_chain_spec(
+					&mut chain_spec_json,
+					&fs::read(runtime.as_path())
+						.map_err(|e| format!("Wasm blob file could not be read: {e}"))?[..],
+					block_height,
+				);
+				let chain_spec_json = serde_json::to_string_pretty(&chain_spec_json)
+					.map_err(|e| format!("to pretty failed: {e}"))?;
+				fs::write(chain_spec_path, chain_spec_json).map_err(|err| err.to_string())?;
+			},
+			ChainSpecBuilderCmd::ConvertToRaw(ConvertToRawCmd { ref input_chain_spec }) => {
+				let chain_spec = ChainSpec::from_json_file(input_chain_spec.clone())?;
+
+				let mut genesis_json =
+					serde_json::from_str::<serde_json::Value>(&chain_spec.as_json(true)?)
+						.map_err(|e| format!("Conversion to json failed: {e}"))?;
+
+				// We want to extract only raw genesis ("genesis::raw" key), and apply it as a patch
+				// for the original json file. However, the file also contains origianl plain
+				// genesis ("genesis::runtimeGenesis") so set it to null so the patch will erase it.
+				genesis_json.as_object_mut().map(|map| {
+					map.retain(|key, _| key == "genesis");
+					map.get_mut("genesis").map(|genesis| {
+						genesis.as_object_mut().map(|genesis_map| {
+							genesis_map
+								.insert("runtimeGenesis".to_string(), serde_json::Value::Null);
+						});
+					});
+				});
+
+				let mut org_chain_spec_json = extract_chain_spec_json(input_chain_spec.as_path())?;
+				json_patch::merge(&mut org_chain_spec_json, genesis_json);
+
+				let chain_spec_json = serde_json::to_string_pretty(&org_chain_spec_json)
+					.map_err(|e| format!("Conversion to pretty failed: {e}"))?;
+				fs::write(chain_spec_path, chain_spec_json).map_err(|err| err.to_string())?;
+			},
+			ChainSpecBuilderCmd::Verify(VerifyCmd { ref input_chain_spec }) => {
+				let chain_spec = ChainSpec::from_json_file(input_chain_spec.clone())?;
+				let _ = serde_json::from_str::<serde_json::Value>(&chain_spec.as_json(true)?)
+					.map_err(|e| format!("Conversion to json failed: {e}"))?;
+			},
+			ChainSpecBuilderCmd::ListPresets(ListPresetsCmd { runtime }) => {
+				let code = fs::read(runtime.as_path())
+					.map_err(|e| format!("wasm blob shall be readable {e}"))?;
+				let caller: GenesisConfigBuilderRuntimeCaller =
+					GenesisConfigBuilderRuntimeCaller::new(&code[..]);
+				let presets = caller
+					.preset_names()
+					.map_err(|e| format!("getting default config from runtime should work: {e}"))?;
+				let presets: Vec<String> = presets
+					.into_iter()
+					.map(|preset| {
+						String::from(
+							TryInto::<&str>::try_into(&preset)
+								.unwrap_or_else(|_| "cannot display preset id")
+								.to_string(),
+						)
+					})
+					.collect();
+				println!("{}", serde_json::json!({"presets":presets}).to_string());
```

Comment:
Nit: I am not sure if this should be a json. Intuitively, I would expect that a cli tool just returns the name in a plain list. But its more a gut feeling, no hard opinion.